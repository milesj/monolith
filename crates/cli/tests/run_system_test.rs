use moon_config::{RunnerConfig, WorkspaceConfig, WorkspaceProjects};
use moon_test_utils::{
    assert_snapshot, create_sandbox_with_config, predicates::prelude::*, Sandbox,
};
use moon_utils::string_vec;
use rustc_hash::FxHashMap;

fn system_sandbox() -> Sandbox {
    let workspace_config = WorkspaceConfig {
        projects: WorkspaceProjects::Sources(FxHashMap::from_iter([
            ("unix".to_owned(), "unix".to_owned()),
            ("windows".to_owned(), "windows".to_owned()),
        ])),
        runner: RunnerConfig {
            // Avoid these in hashes or snapshots
            implicit_inputs: string_vec![],
            ..RunnerConfig::default()
        },
        ..WorkspaceConfig::default()
    };

    let sandbox = create_sandbox_with_config("system", Some(&workspace_config), None, None);

    sandbox.enable_git();
    sandbox
}

#[cfg(not(windows))]
mod unix {
    use super::*;

    #[test]
    fn handles_echo() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("unix:echo");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn handles_ls() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("unix:ls");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn runs_bash_script() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("unix:bash");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn handles_process_exit_zero() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("unix:exitZero");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn handles_process_exit_nonzero() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("unix:exitNonZero");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn passes_args_through() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run")
                .arg("unix:passthroughArgs")
                .arg("--")
                .arg("-aBc")
                .arg("--opt")
                .arg("value")
                .arg("--optCamel=value")
                .arg("foo")
                .arg("'bar baz'")
                .arg("--opt-kebab")
                .arg("123");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn sets_env_vars() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("unix:envVars");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn inherits_moon_env_vars() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("unix:envVarsMoon");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn runs_from_project_root() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("unix:runFromProject");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn runs_from_workspace_root() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("unix:runFromWorkspace");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn retries_on_failure_till_count() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("unix:retryCount");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn can_run_many_targets() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run")
                .arg("unix:foo")
                .arg("unix:bar")
                .arg("unix:baz");
        });

        let output = assert.output();

        assert!(predicate::str::contains("unix:foo | foo").eval(&output));
        assert!(predicate::str::contains("unix:bar | bar").eval(&output));
        assert!(predicate::str::contains("unix:baz | baz").eval(&output));
    }

    mod caching {
        use super::*;
        use moon_cache::RunTargetState;
        use std::fs;

        #[test]
        fn uses_cache_on_subsequent_runs() {
            let sandbox = system_sandbox();

            let assert = sandbox.run_moon(|cmd| {
                cmd.arg("run").arg("unix:outputs");
            });

            assert_snapshot!(assert.output());

            let assert = sandbox.run_moon(|cmd| {
                cmd.arg("run").arg("unix:outputs");
            });

            assert_snapshot!(assert.output());
        }

        #[test]
        fn creates_runfile() {
            let sandbox = system_sandbox();

            sandbox.run_moon(|cmd| {
                cmd.arg("run").arg("unix:outputs");
            });

            assert!(sandbox
                .path()
                .join(".moon/cache/states/unix/runfile.json")
                .exists());
        }

        #[tokio::test]
        async fn creates_run_state_cache() {
            let sandbox = system_sandbox();

            sandbox.run_moon(|cmd| {
                cmd.arg("run").arg("unix:outputs");
            });

            let cache_path = sandbox
                .path()
                .join(".moon/cache/states/unix/outputs/lastRun.json");

            assert!(cache_path.exists());

            let state = RunTargetState::load(cache_path, 0).await.unwrap();

            assert_snapshot!(fs::read_to_string(
                sandbox
                    .path()
                    .join(format!(".moon/cache/hashes/{}.json", state.hash))
            )
            .unwrap());

            assert!(sandbox
                .path()
                .join(".moon/cache/outputs")
                .join(format!("{}.tar.gz", state.hash))
                .exists());
            assert!(sandbox
                .path()
                .join(".moon/cache/states/unix/outputs/stdout.log")
                .exists());
            assert!(sandbox
                .path()
                .join(".moon/cache/states/unix/outputs/stderr.log")
                .exists());
        }
    }

    mod affected_files {
        use super::*;

        #[test]
        fn uses_dot_when_not_affected() {
            let sandbox = system_sandbox();

            let assert = sandbox.run_moon(|cmd| {
                cmd.arg("run").arg("unix:affectedFiles");
            });
            let output = assert.output();

            assert!(predicate::str::contains("Args: .\n").eval(&output));
        }

        #[test]
        fn uses_rel_paths_when_affected() {
            let sandbox = system_sandbox();

            sandbox.create_file("unix/input1.txt", "");
            sandbox.create_file("unix/input2.txt", "");

            let assert = sandbox.run_moon(|cmd| {
                cmd.arg("run").arg("unix:affectedFiles").arg("--affected");
            });
            let output = assert.output();

            assert!(predicate::str::contains("Args: ./input1.txt ./input2.txt").eval(&output));
        }

        #[test]
        fn sets_env_var() {
            let sandbox = system_sandbox();

            sandbox.create_file("unix/input1.txt", "");
            sandbox.create_file("unix/input2.txt", "");

            let assert = sandbox.run_moon(|cmd| {
                cmd.arg("run")
                    .arg("unix:affectedFilesEnvVar")
                    .arg("--affected");
            });
            let output = assert.output();

            assert!(
                predicate::str::contains("MOON_AFFECTED_FILES=./input1.txt,./input2.txt")
                    .eval(&output)
            );
        }
    }
}

#[cfg(windows)]
mod system_windows {
    use super::*;

    #[test]
    fn runs_bat_script() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("windows:bat");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn handles_process_exit_zero() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("windows:exitZero");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn handles_process_exit_nonzero() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("windows:exitNonZero");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn passes_args_through() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run")
                .arg("windows:passthroughArgs")
                .arg("--")
                .arg("-aBc")
                .arg("--opt")
                .arg("value")
                .arg("--optCamel=value")
                .arg("foo")
                .arg("'bar baz'")
                .arg("--opt-kebab")
                .arg("123");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn sets_env_vars() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("windows:envVars");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn inherits_moon_env_vars() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("windows:envVarsMoon");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn runs_from_project_root() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("windows:runFromProject");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn runs_from_workspace_root() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("windows:runFromWorkspace");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn retries_on_failure_till_count() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run").arg("windows:retryCount");
        });

        assert_snapshot!(assert.output());
    }

    #[test]
    fn can_run_many_targets() {
        let sandbox = system_sandbox();

        let assert = sandbox.run_moon(|cmd| {
            cmd.arg("run")
                .arg("windows:foo")
                .arg("windows:bar")
                .arg("windows:baz");
        });

        let output = assert.output();

        assert!(predicate::str::contains("windows:foo | foo").eval(&output));
        assert!(predicate::str::contains("windows:bar | bar").eval(&output));
        assert!(predicate::str::contains("windows:baz | baz").eval(&output));
    }

    mod caching {
        use super::*;
        use moon_cache::RunTargetState;
        use std::fs;

        #[test]
        fn uses_cache_on_subsequent_runs() {
            let sandbox = system_sandbox();

            let assert = sandbox.run_moon(|cmd| {
                cmd.arg("run").arg("windows:outputs");
            });

            assert_snapshot!(assert.output());

            let assert = sandbox.run_moon(|cmd| {
                cmd.arg("run").arg("windows:outputs");
            });

            assert_snapshot!(assert.output());
        }

        #[test]
        fn creates_runfile() {
            let sandbox = system_sandbox();

            sandbox.run_moon(|cmd| {
                cmd.arg("run").arg("windows:outputs");
            });

            assert!(sandbox
                .path()
                .join(".moon/cache/states/windows/runfile.json")
                .exists());
        }

        #[tokio::test]
        async fn creates_run_state_cache() {
            let sandbox = system_sandbox();

            sandbox.run_moon(|cmd| {
                cmd.arg("run").arg("windows:outputs");
            });

            let cache_path = sandbox
                .path()
                .join(".moon/cache/states/windows/outputs/lastRun.json");

            assert!(cache_path.exists());

            let state = RunTargetState::load(cache_path, 0).await.unwrap();

            assert_snapshot!(fs::read_to_string(
                sandbox
                    .path()
                    .join(format!(".moon/cache/hashes/{}.json", state.hash))
            )
            .unwrap());

            assert!(sandbox
                .path()
                .join(".moon/cache/outputs")
                .join(format!("{}.tar.gz", state.hash))
                .exists());
            assert!(sandbox
                .path()
                .join(".moon/cache/states/windows/outputs/stdout.log")
                .exists());
            assert!(sandbox
                .path()
                .join(".moon/cache/states/windows/outputs/stderr.log")
                .exists());
        }
    }
}
