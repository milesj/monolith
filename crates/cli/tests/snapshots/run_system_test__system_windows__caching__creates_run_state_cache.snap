---
source: crates/cli/tests/run_system_test.rs
expression: "fs::read_to_string(fixture.path().join(format!(\".moon/cache/hashes/{}.json\",\n                state.hash))).unwrap()"
---
[
  {
    "command": "cmd.exe",
    "args": [
      "./outputs.bat",
      "/c",
      "/q"
    ],
    "deps": {},
    "envVars": {},
    "inputs": {
      "windows/cwd.bat": "f97d5bfb654f63a59a4158e4969707c8e1c29ece",
      "windows/echo.bat": "cf9137992257ec89c223a5d866458a94a361e0fd",
      "windows/envVars.bat": "af068672f8e5f2ca1aa1b5fec93df8fb253bdd0c",
      "windows/envVarsMoon.bat": "f378867c19344c38e6587248c9a6b1643b49bbca",
      "windows/exitNonZero.bat": "8204ab8df26ed8af772def3d3701a0df93d81ebe",
      "windows/exitZero.bat": "0cc40242e00650f6cf05f8c523f5e9afb268fb1f",
      "windows/moon.yml": "4ed9b777bf245c05de077d64f4e82812d86c6dcc",
      "windows/outputs.bat": "df7bd8b019bf17a024f17dadcff7e9b2cdfdb471",
      "windows/passthroughArgs.bat": "f00267f81c05165ee19da2c5940bc06e39217b61",
      "windows/standard.bat": "3cbc31ab39b138c232a071378b428ff015de8dae"
    },
    "outputs": [
      "file.txt",
      "folder"
    ],
    "projectDeps": [],
    "target": "windows:outputs",
    "version": "1"
  },
  {
    "version": "1"
  }
]
